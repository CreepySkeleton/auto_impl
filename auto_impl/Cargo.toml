[package]
name = "auto_impl"
version = "0.1.0"
authors = ["Ashley Mannix <ashleymannix@live.com.au>"]

[lib]
proc-macro = true

[dependencies]
quote = "*"
auto_impl_internals = { version = "*", path = "../auto_impl_internals" }