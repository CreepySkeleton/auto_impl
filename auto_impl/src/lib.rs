#![feature(proc_macro)]

extern crate quote;
extern crate proc_macro;
extern crate auto_impl_internals;

use std::str::FromStr;
use quote::Tokens;
use proc_macro::TokenStream;

fn auto_impl_expand(attrs: TokenStream, proc_tokens: TokenStream) -> Result<TokenStream, String> {
    let mut tokens = Tokens::new();
    tokens.append(proc_tokens.to_string());

    let mut attr_tokens = Tokens::new();
    attr_tokens.append(format!("#[auto_impl{}]", attrs.to_string()));

    let impl_types = derive_internals::parse_impl_types(attr_tokens)?;

    let tokens = derive_internals::auto_impl_expand(&impl_types, tokens)?;

    TokenStream::from_str(&tokens.to_string()).map_err(|e| format!("{:?}", e))
}

#[proc_macro_attribute]
pub fn auto_impl(attrs: TokenStream, proc_tokens: TokenStream) -> TokenStream {
    match auto_impl_expand(attrs, proc_tokens) {
        Ok(tokens) => tokens,
        Err(e) => panic!("auto impl error: {}", e)
    }
}
