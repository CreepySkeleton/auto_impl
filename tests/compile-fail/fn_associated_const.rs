
extern crate auto_impl;
use auto_impl::auto_impl;


#[auto_impl(Fn)]
trait Foo {
    const LEN: usize;

    fn a(&self);
}
