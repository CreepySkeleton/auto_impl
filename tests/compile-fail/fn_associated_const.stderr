error: this trait cannot be auto-implemented for Fn-traits (only traits with exactly one method and no other items are allowed)

note: auto-impl requested here
 --> $DIR/fn_associated_const.rs:5:1
  |
5 | trait Foo {
  | ^^^^^
